builder/builder.go

```diff
@@ -201,83 +201,58 @@ func (b *Builder) SignManifestMoM() error {
 	return nil
 }
 
-// UpdateRepo will fetch the clr-bundles for our configured Clear Linux version
-func (b *Builder) UpdateRepo(ver string, allbundles bool) {
-	// Make the folder to store all clr-bundles version
-	if _, err := os.Stat("clr-bundles"); err != nil {
-		if err = os.Mkdir("clr-bundles", 0777); err != nil {
-			fmt.Fprintf(os.Stderr, "ERROR: Failed to create clr-bundles: %s\n", err.Error())
-		}
-	}
+const (
+	upstreamBundlesBaseDir   = ".mixer/upstream-bundles/"
+	upstreamBundlesVerDir    = "clr-bundles-%s"
+	upstreamBundlesBundleDir = "bundles"
+)
 
-	repo := "clr-bundles/clr-bundles-" + ver + ".tar.gz"
-	if _, err := os.Stat(repo); err == nil {
-		fmt.Println("Already downloaded " + repo)
-		return
-	}
+func getUpstreamBundlesDir(ver string) string {
+	return filepath.Join(upstreamBundlesBaseDir, fmt.Sprintf(upstreamBundlesVerDir, ver), upstreamBundlesBundleDir)
+}
 
-	URL := "https://github.com/clearlinux/clr-bundles/archive/" + ver + ".tar.gz"
-	err := helpers.Download(repo, URL)
-	if err != nil {
-		fmt.Fprintf(os.Stderr, "ERROR: Failed to download clr-bundles, make sure the version is valid: %s\n", err)
-		os.Exit(1)
+func (b *Builder) getUpstreamBundles(ver string, prune bool) error {
+	// Make the folder to store upstream bundles if does not exist
+	if _, err := os.Stat(upstreamBundlesBaseDir); os.IsNotExist(err) {
+		if err = os.MkdirAll(upstreamBundlesBaseDir, 0777); err != nil {
+			return errors.Wrap(err, "Failed to create upstream-bundles dir.")
+		}
 	}
 
-	// FIXME: Maybe use Go's tar or compress packages to do this
-	_, err = exec.Command("tar", "-xzf", repo, "-C", "clr-bundles/").Output()
-	if err != nil {
-		helpers.PrintError(err)
-		os.Exit(1)
-	}
-	bundles := b.Bundledir
-	if _, err := os.Stat(bundles); os.IsNotExist(err) {
-		clrbundles := "clr-bundles/clr-bundles-" + ver + "/bundles/"
-		if err = os.Mkdir(bundles, 0777); err != nil {
-			helpers.PrintError(err)
-			os.Exit(1)
+	bundleDir := getUpstreamBundlesDir(ver)
+
+	// Clear out other bundle dirs if needed
+	if prune {
+		files, err := ioutil.ReadDir(upstreamBundlesBaseDir)
+		if err != nil {
+			return errors.Wrap(err, "Failed to read bundles for pruning.")
 		}
-		// Copy all bundles over into mix-bundles if -all passed
-		if allbundles == true {
-			files, err := ioutil.ReadDir("clr-bundles/clr-bundles-" + ver + "/bundles/")
-			if err != nil {
-				helpers.PrintError(err)
-				os.Exit(1)
-			}
-			for _, file := range files {
-				if err = helpers.CopyFile(bundles+"/"+file.Name(), clrbundles+file.Name()); err != nil {
-					helpers.PrintError(err)
+		for _, file := range files {
+			// Skip the current version cache, but delete all others
+			if strings.Compare(file.Name(), fmt.Sprintf(upstreamBundlesVerDir, ver)) != 0 {
```

Comment:
Use ```if a != b``` for comparing the strings.

Optional: save the result of fmt.Sprintf in a local variable and reuse it.

Reply:
>Use if a != b for comparing the strings.

Just read the [comments in strings.Compare source](https://golang.org/src/strings/compare.go). Good to know.

>Optional: save the result of fmt.Sprintf in a local variable and reuse it.

I might be missing your point, but I'm not re-calculating that value anywhere in this function, so not sure what this would do.

Reply:
Optional: the suggestion was outside the for loop, you create a variable (e.g. 'verDir') that holds the result of fmt.Sprintf(...) call. Since that will not change for each iteration, you can reuse that for each iteration.

Reply:
Yeah, that's good practice. Kind of gotten out of habit for things like that, as most compilers for other languages will optimize that stuff automatically these days; does Go not?

Reply:
My guess in this case is that Go compiler can't infer that Sprintf will not have side effects (unless it inlines it, which I don't think happen in this case), so it can't automatically make this transformation. (We should just check compiler output!)

Note: in this case I think the change is worth even if it doesn't optimize anything.

Reply:
Fixed in forthcoming push.

Reply:
@kevincwells Actually few compilers will do that kind of optimization of function calls. In particular gcc C will not unless the function being called is annotated either `pure` or `const`. Of course some languages are side effect free and compilers for those languages will do the code hoisting you mention.

Reply:
@icarus-sparry & @cmarcelo -- yeah, you're both right. I'm running on little sleep today, so I wasn't thinking about function side effects when I made my remark.